src/Features/Core/Portable/MetadataAsSource/FileSystemMetadataDocumentPersister.cs

```diff
@@ -0,0 +1,177 @@
+            : SpecializedTasks.Null<SourceText>();
+    }
+    public async Task<bool> WriteMetadataDocumentAsync(string documentPath, Encoding encoding, SourceText text, Action<Exception>? logFailure, CancellationToken cancellationToken)
+    {
+        try
+        {
+            var directoryName = Path.GetDirectoryName(documentPath)!;
+            if (!await EnsureDirectoryExistsAsync(directoryName, cancellationToken).ConfigureAwait(false))
+                throw new IOException($"Unable to create directory: {directoryName}");
+
+            using (var textWriter = new StreamWriter(documentPath, append: false, encoding))
+            {
+                text.Write(textWriter, cancellationToken);
+            }
+
+            // Mark read-only
+            new FileInfo(documentPath).IsReadOnly = true;
+            return true;
+        }
+        catch (Exception ex) when (IOUtilities.IsNormalIOException(ex))
+        {
+            // If we hit an IO exception, we assume the file could not be written.
+            logFailure?.Invoke(ex);
+            return false;
+        }
+    }
+
+    public string GenerateDocumentPath(Guid identifier, string providerName, string fileName)
+    {
+        return Path.Combine(_rootTemporaryPathWithGuid, providerName, identifier.ToString(), fileName);
+    }
+
+    public string ConvertFilePathToDocumentPath(Guid identifier, string providerName, string filePath)
+    {
+        // We already have a file path, and this persister operates on files.  No need to do anything to it.
+        return filePath;
+    }
+
+    public void CleanupGeneratedDocuments()
+    {
+        try
+        {
+            if (Directory.Exists(_rootTemporaryPath))
+            {
+                // Let's look through directories to delete.
+                foreach (var directoryInfo in new DirectoryInfo(_rootTemporaryPath).EnumerateDirectories())
+                {
+                    // Is there a mutex for this one?  If so, that means it's a folder open in another VS instance.
+                    // We should leave it alone.  If not, then it's a folder from a previous VS run.  Delete that
+                    // now.
+                    if (Mutex.TryOpenExisting(CreateMutexName(directoryInfo.Name), out var acquiredMutex))
+                    {
+                        acquiredMutex.Dispose();
+                    }
+                    else
+                    {
+                        TryDeleteFolderWhichContainsReadOnlyFiles(directoryInfo.FullName);
+                    }
+                }
+            }
+        }
+        catch (Exception)
+        {
+        }
+
+        static void TryDeleteFolderWhichContainsReadOnlyFiles(string directoryPath)
+        {
+            try
+            {
+                foreach (var fileInfo in new DirectoryInfo(directoryPath).EnumerateFiles("*", SearchOption.AllDirectories))
+                    IOUtilities.PerformIO(() => fileInfo.IsReadOnly = false);
```

Comment:
curious why we bother with this (but don't care that much).

Reply:
Not 100% sure, but this was unchanged from what it was doing before.

Reply:
Delete on a read only file can throw.